gaia-header.js

```diff
@@ -34,124 +54,519 @@ module.exports = Component.register('gaia-header', {
+   * avoid reading from the DOM when
+   * it may not be loaded yet.
+   *
    * @private
    */
   attached: function() {
-    this.rerunFontFit();
+    debug('attached');
+    this.runFontFitSoon();
+    this.observerStart();
   },
 
   /**
-   * Called when one of the attributes
-   * on the element changes.
+   * Called when the element is
+   * detached from the DOM.
    *
    * @private
    */
-  attributeChanged: function(attr) {
-    if (attr === 'action') {
-      this.configureActionButton();
-      this.rerunFontFit();
-    }
+  detached: function() {
+    debug('detached');
+    this.observerStop();
+    this.clearPending();
   },
 
   /**
-   * Runs the logic to size and position
-   * header text inside the available space.
+   * Clears pending `.nextTick()`s
    *
    * @private
    */
-  runFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
-      fontFit.observeHeadingChanges(this.els.headings[i]);
+  clearPending: function() {
+    for (var key in this.pending) {
+      this.pending[key].clear();
+      delete this.pending[key];
     }
   },
 
   /**
-   * Rerun font-fit logic.
+   * Run the font-fit logic and
+   * center the title.
+   *
+   * The styles are calculated synchronously,
+   * but then set asynchronously. This means
+   * this function can be hammered in one turn,
+   * and the title styles will only be written
+   * once, with the latest styles.
+   *
+   * @return {Promise}
+   * @public
+   */
+  runFontFit: function() {
+    debug('run font-fit');
+
+    // Nothing is run if `no-font-fit` attr is present
+    if (this.noFontFit) { return Promise.reject('no-font-fit'); }
+
+    var titles = this.els.titles;
+    var space = this.getTitleSpace();
+    var styles = [].map.call(titles, (el) => this.getTitleStyle(el, space));
+
+    // Update the title styles using the latest
+    // styles. This function can be called many
+    // times but will only run once in each tick.
+    return this.setTitleStylesSoon(styles);
+  },
+
+  /**
+   * Microtask debounced `runFontFit`
    *
-   * TODO: We really need an official API for this.
+   * @private
+   */
+  runFontFitSoon: function() {
+    debug('run font-fit soon');
+    if (this.pending.runFontFitSoon) { return; }
+    this.pending.runFontFitSoon = this.nextTick(() => {
+      delete this.pending.runFontFitSoon;
+      this.runFontFit();
+    });
+  },
+
+  /**
+   * Get the style properties required
+   * to fit and position the title text.
    *
+   * @param  {HTMLH1Element} el
+   * @param  {Number} space
+   * @return {Object} {fontSize, marginStart, overflowing, id}
    * @private
    */
-  rerunFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
+  getTitleStyle: function(el, space) {
+    debug('get el style', el, space);
+    var text = el.textContent;
+    var styleId = text + space;
```

Comment:
it would be technically more correct to use `JSON.stringify({text: text, space: space});` but because it's likely more costly we can keep the simple concatenation we have here.